feat(typedoc-plugin-appium): extract descriptions of return values from builtins

boneskull opened the pull request:
If a command is not documented with a `@returns` tag, but a builtin method _does_ have this tag (e.g., `ExternalDriver.createSession()`), then the plugin will use the tag from the builtin.

If a command _does_ have a `@returns` tag, but it provides no description, then the plugin will use the original `@returns` tag (which may have a type) and combine it with the content/description from the builtin (if any).

This requires further inspection of `SignatureReflection` data, which needed a new set of "comment finders", which prompted me to refactor the comment finders into their own files (because it was becoming unwieldy).  The call to `deriveComment` happens in the static `BaseCommandData.rewriteSignature()` method (whereas before it was just a straightforward clone operation).

- Fixed the block tag aggregation to prefer contentful over contentless tags in all cases.
- Modified the `@returns` tag of `FakeDriver.createSession()` for testing
- Moved some types around that were used outside of their sibling modules (e.g. `CommentSource` which used to be called `CommentSourceType`)
- pruned dead imports

Closes #18196

boneskull committed:
feat(typedoc-plugin-appium): extract descriptions of return values from builtins

If a command is not documented with a `@returns` tag, but a builtin method _does_ have this tag (e.g., `ExternalDriver.createSession()`), then the plugin will use the tag from the builtin.

If a command _does_ have a `@returns` tag, but it provides no description, then the plugin will use the original `@returns` tag (which may have a type) and combine it with the content/description from the builtin (if any).

This requires further inspection of `SignatureReflection` data, which needed a new set of "comment finders", which prompted me to refactor the comment finders into their own files (because it was becoming unwieldy).  The call to `deriveComment` happens in the static `BaseCommandData.rewriteSignature()` method (whereas before it was just a straightforward clone operation).

- Fixed the block tag aggregation to prefer contentful over contentless tags in all cases.
- Modified the `@returns` tag of `FakeDriver.createSession()` for testing
- Moved some types around that were used outside of their sibling modules (e.g. `CommentSource` which used to be called `CommentSourceType`)
- pruned dead imports

boneskull commented:
Current dependencies on/for this PR:
* master
  * **PR #18202** <a href="https://app.graphite.dev/github/pr/appium/appium/18202" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/appium/appium/18202?utm_source=stack-comment).

boneskull commented:
Here, the text at the very bottom was extracted from the docstring in `@appium/types`: https://github.com/appium/appium/blob/64554b967aa6d48864505d03b9c2e57b167059a8/packages/types/lib/driver.ts#L404-L426

![image](https://user-images.githubusercontent.com/924465/218884606-8a397b7f-c434-41f3-b573-3da325c70fc3.png)

jlipps commented on packages/typedoc-plugin-appium/lib/converter/comment/index.ts:
```diff
@@ -0,0 +1,182 @@
+      const tag = tags.find((t) => t.content.length) ?? tags[0];
+      finalBlockTags.push(tag);
+    } else {
+      // should not happen
```
do we want to throw an error here so if it _does_ happen we know that it happened?

boneskull replied:
(I'll log it)

jlipps commented on packages/typedoc-plugin-appium/lib/converter/comment/method.ts:
```diff
@@ -0,0 +1,79 @@
+      //
+      // after looping thru the finders, if we have a comment in the list of `commentData`
+      // objects, return the first one found.
+      const comment = MethodCommentFinders.filter(
```
probably easier to read if you put `.filter` on the line below so it lines up with `.map` and `.find`

boneskull replied:
prettier won't let me

jlipps commented on packages/typedoc-plugin-appium/lib/converter/comment/types.ts:
```diff
@@ -0,0 +1,31 @@
+
+/**
+ * A function which queries some object for a {@linkcode Comment}, and if found,
+ * its {@linkcode Comment.summary summary}.
```
should this be `{@linkcode Comment.summary} summary`?

boneskull replied:
it does the same thing:

![image](https://user-images.githubusercontent.com/924465/219764319-995669e0-5a8c-4ecd-850d-f86f1db5d9bb.png)

jlipps commented on packages/typedoc-plugin-appium/lib/converter/types.ts:
```diff
@@ -276,3 +277,87 @@ export type ExecuteMethodCommandReflection = CommandReflection & {
+  rb: 'Ruby',
+  ruby: 'Ruby',
+  java: 'Java',
+}) satisfies Record<string, string>;
```
👏

boneskull committed:
chore(typedoc-plugin-appium): add a fallback logger
due to the architecture of TypeDoc, it is not straightforward for a random module in a plugin to have access to the logger of the current `Application`.  This change abuses TypeDoc to enable logging for random modules.

boneskull commented:
@jlipps If CI passes, I'm going to merge this, but please look at https://github.com/appium/appium/pull/18202/commits/be84cfb0aae305a78b8b7bba8ccd4199b994e584 if you get a chance, because it is shameful
